### PR TITLE
[ci] Compile BCL with interpreter

### PR DESCRIPTION
**Pull request by lewurm**
Fixes https://github.com/mono/mono/issues/7418

### CODE REVIEW

File: mcs/build/config-default.make [hunk 1]
```diff
@@ -9,7 +9,11 @@
 # Use UTF-8 as the default encoding for compilers
 CODEPAGE = 65001
 
+ifdef TEST_WITH_INTERPRETER
+RUNTIME_FLAGS = --interpreter
+else
 RUNTIME_FLAGS =
+endif
```
Comment (akoeplinger) [inline mcs/build/config-default.make]: I'd move it to the TEST_WITH_INTERPRETER if clause we have already right below.

Comment (luhenry) [reply mcs/build/config-default.make]: +1

**Comment by luhenry** [comment]
@lewurm why `do-not-merge`? I would love to see it happen sooner rather than later, and it won't impact any of our customers, just CI

**Comment by BrzVlad** [comment]
@lewurm if it builds fine, I think it would be nice try it a few more times before merging,  just in case, since this would be a required lane.

**Comment by akoeplinger** [comment]
@luhenry the `do-not-merge` label is fine, it shows that this is still experimental so someone (like me 😄) doesn't accidentally merge it

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by lewurm** [comment]
@marek-safar there is still an mcs issue on `sizeof-empty-structs.cs`:
```
sizeof-empty-structs.cs(87,28): error CS1929: Type `System.Collections.Generic.List<(string, int)>' does not contain a member `SequenceEqual' and the best extension method overload 
Compilation failed: 1 error(s), 0 warnings
```

**Comment by luhenry** [comment]
Some progress: we now can run testinterp, and the list of failing tests are `bug-60848.exe`, `DebuggerTests.LineNumbers`, `DebuggerTests.SetIP, `DebuggerTests.SetIPSingleStep, `DebuggerTests.SingleStepping`. They all fail on all platforms so I would assume they are related to the change to mcs.

@marek-safar @kumpera if we want to even just avoid regressions with MCS we should run the tests at least with the JIT (on top of the build-only CI check we have nowadays).

**Comment by marek-safar** [comment]
The debugger test needs mcs specific changes.

We could build runtime tests with mcs but the failures are from new tests so no regressions found and it'll block us from using any new C# syntax mcs does not support there

**Comment by marek-safar** [comment]
>For the debugger tests, that means that we don't support the debugger with code generated by MCS in general then.

We do but as generated IL is different the debugger verification checks are different because the tests are quite low level

>For the runtime tests, we already have a status check for Linux x64 mcs which compiles all BCL, BCL tests and runtimes tests, and it's a required status check, meaning any breakage shouldn't happen.

Incorrect, we don't build runtime tests with mcs in `Linux x64 mcs`

>So the question bilds down to: how much do we care about MCS? Is it still a first-class compiler, in this case it should work just as well with MCS or CSC; or is it a second-class compiler?

mcs is a fallback compiler right now, no reason to fight for first-class vs second-class.

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by lewurm** [comment]
@monojenkins build Linux x64 Interpreter

**Comment by marek-safar** [comment]
@monojenkins build failed

**Comment by luhenry** [comment]
@monojenkins build failed

**Pull request closed without merging**